fix: Don't use copyStyleSheets with documentPIP

beaufortfrancois opened the pull request:
Document Picture-in-Picture API owners are [removing `copyStyleSheets` option](https://github.com/WICG/document-picture-in-picture/pull/79) as it can be done in JavaScript. This PR removes this option and copy all style sheets internally to the PiP window.

@mister-ben Please review.

mister-ben commented on src/js/player.js:
```diff
@@ -3109,6 +3109,36 @@ class Player extends Component {
+   *
+   * @private
+   */
+  copyStyleSheetsToWindow_(win) {
```
This is probably better in src/js/utils/dom.js than exposed as a method on the player.

beaufortfrancois replied:
Done.

mister-ben reviewed (changes requested):
Thanks for the update. I'm not really sure I agree that it makes any more sense for each player library to manage copying stylesheets, but if that's how it is...

The copy method would ideally not be exposed on the player, as the inline comment. We'll need to address test coverage too.

beaufortfrancois commented:
> The copy method would ideally not be exposed on the player, as the inline comment. We'll need to address test coverage too.

I've moved it and added some basic tests. Let me know if that looks good to you.

beaufortfrancois commented:
(gentle ping)

mister-ben commented:
@beaufortfrancois Tahnks for the updates. I've been out for a couple of weeks, back now and will check this week.

mister-ben commented on src/js/utils/dom.js:
```diff
@@ -857,3 +857,33 @@ export function computedStyle(el, prop) {
+      try {
+        return [...styleSheet.cssRules].map((rule) => rule.cssText).join('');
+      } catch (e) {
+        const link = document.createElement('link');
```
When would it be expected for code in the try block to fail so we reach here? Running locally and in CI the concatenation of styles works for both `<style>` and `<link>` els, hence the test failures - no `<link>` el is added to the doc.

If `<links>` did need to be added here, they would being added in the order they occur but the content of all `<style>`s would be in a `<style>` after those `<link>`s, potentially changing the original order of rules.

beaufortfrancois replied:
Accessing `cssRules` from cross-origin `<link>` elements such as `<link href="https://cdn.com/styles.css" rel="stylesheet">` raise an error: "Failed to read the 'cssRules' property from 'CSSStyleSheet': Cannot access rules". That's why we try/catch here and append them manually instead.

FWIW, tests pass locally.

mister-ben replied:
The reason the tests fail is that adding the the `<link>` stylesheet is async and it's not present in `document.stylesSheets` in time for the test.

mister-ben commented on src/js/utils/dom.js:
```diff
@@ -857,3 +857,33 @@ export function computedStyle(el, prop) {
+ *
+ */
+export function copyStyleSheetsToWindow(win) {
+  const allCSS = [...document.styleSheets]
+    .map((styleSheet) => {
+      try {
+        return [...styleSheet.cssRules].map((rule) => rule.cssText).join('');
+      } catch (e) {
+        const link = document.createElement('link');
+
+        link.rel = 'stylesheet';
+        link.type = styleSheet.type;
+        link.media = styleSheet.media;
+        link.href = styleSheet.href;
+        win.document.head.appendChild(link);
+      }
+    })
+    .filter(Boolean)
+    .join('\n');
+  const style = document.createElement('style');
+
+  style.textContent = allCSS;
+  win.document.head.appendChild(style);
```
Adding each stylesheet in the original order would avoid changing the preceence of rules.

```suggestion
  [...document.styleSheets].forEach((styleSheet) => {
    try {
      const cssRules = [...styleSheet.cssRules].map((rule) => rule.cssText).join('');
      const style = document.createElement('style');

      style.textContent = cssRules;
      win.document.head.appendChild(style);
    } catch (e) {
      const link = document.createElement('link');

      link.rel = 'stylesheet';
      link.type = styleSheet.type;
      link.media = styleSheet.media;
      link.href = styleSheet.href;
      win.document.head.appendChild(link);
    }
  });
```

mister-ben commented on test/unit/utils/dom.test.js:
```diff
@@ -686,3 +686,48 @@ QUnit.test('isSingleLeftClick() checks return values for mousedown event', funct
 });
+
+QUnit.test('Dom.copyStyleSheetsToWindow() copies all style sheets to a window', function(assert) {
+  const fakeWindow = document.createElement('div');
```
This tests would need to use async to wait for the <link> el.

```suggestion
  const fakeWindow = document.createElement('div');
  const done = assert.async();

  assert.expect(7);
```

mister-ben commented on test/unit/utils/dom.test.js:
```diff
@@ -686,3 +686,48 @@ QUnit.test('isSingleLeftClick() checks return values for mousedown event', funct
+  link.type = 'text/css';
+  link.media = 'print';
+  link.href = 'http://asdf.com/styles.css';
+  document.head.appendChild(link);
+
+  Dom.copyStyleSheetsToWindow(fakeWindow);
+
+  assert.expect(7);
+
+  assert.strictEqual(fakeWindow.document.head.querySelectorAll('style').length, 1, 'the fake window has one <style> element only');
+
+  const fakeWindowStyle = fakeWindow.document.head.querySelectorAll('style')[0];
+
+  assert.true(fakeWindowStyle.textContent.includes(style1.textContent), 'the <style> in the fake window contains content from first <style> element');
+  assert.true(fakeWindowStyle.textContent.includes(style2.textContent), 'the <style> in the fake window contains content from second <style> element');
+
+  assert.strictEqual(fakeWindow.document.head.querySelectorAll('link[rel=stylesheet]').length, 1, 'the fake window has one <link> stylesheet element');
+
+  const fakeWindowLink = fakeWindow.document.head.querySelectorAll('link[rel=stylesheet]')[0];
+
+  assert.strictEqual(fakeWindowLink.type, link.type, 'the <style> type attribute in the fake window is the one from <link> element');
+  assert.strictEqual(fakeWindowLink.href, link.href, 'the <style> href attribute in the fake window is the one from <link> element');
+  assert.strictEqual(fakeWindowLink.media, link.media, 'the <style> media attribute in the fake window is the one from <link> element');
```
Then the tests can be run after the link stylesheet exists in `document.styleSheets`

```suggestion
  const containsRulesFromStyle = (el) => {
    return Boolean([...fakeWindow.document.head.querySelectorAll('style')].find(s => {
      return s.textContent.includes(el.textContent);
    }));
  };

  link.onload = link.onerror = () => {
    Dom.copyStyleSheetsToWindow(fakeWindow);

    assert.strictEqual(fakeWindow.document.head.querySelectorAll('style, link').length, document.styleSheets.length, 'the fake window has as many <style> or <link> elements as document.styleSheets');

    assert.true(containsRulesFromStyle(style1), 'a <style> in the fake window contains content from first <style> element');
    assert.true(containsRulesFromStyle(style2), 'a <style> in the fake window contains content from second <style> element');

    assert.strictEqual(fakeWindow.document.head.querySelectorAll('link[rel=stylesheet]').length, 1, 'the fake window has one <link> stylesheet element');

    const fakeWindowLink = fakeWindow.document.head.querySelectorAll('link[rel=stylesheet]')[0];

    assert.strictEqual(fakeWindowLink.type, link.type, 'the <style> type attribute in the fake window is the one from <link> element');
    assert.strictEqual(fakeWindowLink.href, link.href, 'the <style> href attribute in the fake window is the one from <link> element');
    assert.strictEqual(fakeWindowLink.media, link.media, 'the <style> media attribute in the fake window is the one from <link> element');

    done();
  };

  document.head.appendChild(link);
```

mister-ben commented:
We'll need to do something with this as Chrome Canary already remove `copyStylesheets`. @beaufortfrancois any comments re the suggested changes to make sure the order of stylesheets is preserved?
